plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClusteringPredicatesExtractor.java

```diff
@@ -65,86 +64,112 @@ private static ClusteringPushDownResult getClusteringKeysSet(List<CassandraColum
+                            if (discreteValues.getValuesCount() == 0) {
+                                return null;
+                            }
+                            if (discreteValues.getValuesCount() == 1) {
+                                fullyPushedDomains.put(columnHandle, domain);
+                                return format("%s = %s",
+                                        CassandraCqlUtils.validColumnName(columnHandle.getName()),
+                                        discreteValues.getValues().stream().findFirst()
+                                                .orElseThrow());
+                            }
+                            if (isInStatementNotAllowed(clusteringColumns, cassandraVersion, currentlyProcessedClusteringColumn)) {
+                                return null;
                             }
-                            String predicate = CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN ("
-                                    + Joiner.on(",").join(discreteValuesList.build()) + ")";
-                            return predicate;
+
+                            fullyPushedDomains.put(columnHandle, domain);
+
+                            return discreteValues.getValues().stream()
+                                    .map(columnHandle.getCassandraType()::toCqlLiteral)
+                                    .reduce((first, second) -> first + "," + second)
+                                    .map(inValues -> CassandraCqlUtils.validColumnName(columnHandle.getName()) + " IN (" + inValues + " )")
+                                    .orElse(null);
                         }
                         return null;
                     }, allOrNone -> null);
 
             if (predicateString == null) {
                 break;
             }
-            // IN restriction only on last clustering column for Cassandra version = 2.1
-            if (predicateString.contains(" IN (") && cassandraVersion.compareTo(VersionNumber.parse("2.2.0")) < 0 && currentClusteringColumn != (clusteringColumns.size() - 1)) {
-                break;
-            }
             clusteringColumnSql.add(predicateString);
-            domainsBuilder.put(columnHandle, domain);
             // Check for last clustering column should only be restricted by range condition
             if (predicateString.contains(">") || predicateString.contains("<")) {
                 break;
             }
-            currentClusteringColumn++;
+            allProcesedClusteringColumns++;
         }
         List<String> clusteringColumnPredicates = clusteringColumnSql.build();
 
-        return new ClusteringPushDownResult(domainsBuilder.build(), Joiner.on(" AND ").join(clusteringColumnPredicates));
+        return new ClusteringPushDownResult(fullyPushedDomains.build(), Joiner.on(" AND ").join(clusteringColumnPredicates));
+    }
+
+    /**
+     * IN restriction allowed only on last clustering column for Cassandra version = 2.1
```

Comment:
`<= 2.1` ?

Reply:
Code seems to indicate that. I've just moved it around. Will change the comment.